src/gss_names.c

```diff
@@ -62,36 +62,137 @@ static uint32_t string_split(uint32_t *minor_status, char sep,
     return GSSERR();
 }
 
+/* Form of names allowed in GSSNTLMSSP now:
+ *
+ * Standard Forms:
+ *  foo
+ *      USERNAME: foo
+ *      DOMAIN: <null>
+ *
+ *  BAR\foo
+ *      USERNAME: foo
+ *      DOMAIN: BAR
+ *
+ *  foo@BAR
+ *      USERNAME: foo
+ *      DOMAIN: BAR
+ *
+ * Enterprise name forms:
+ *  foo\@bar.example.com
+ *      USERNAME: foo@bar.example.com
+ *      DOMAIN: <null>
+ *
+ *  foo\@bar.example.com@BAR
+ *      USERNAME: foo@bar.example.com
+ *      DOMAIN: BAR
+ *
+ *  BAR\foo@bar.example.com
+ *      USERNAME: foo@bar.example.com
+ *      DOMAIN: BAR
+ *
+ *  BAR@dom\foo@bar.example.com
+ *      USERNAME: foo@bar.example.com
+ *      DOMAIN: BAR@dom
+ *
+ * Invalid forms:
+ *  BAR@dom\@foo..
+ *  DOM\foo\@bar
+ *  foo@bar\@baz
+ */
 #define MAX_NAME_LEN 1024
 static uint32_t get_enterprise_name(uint32_t *minor_status,
                                     const char *str, size_t len,
-                                    char **username)
+                                    char **domain, char **username)
 {
     uint32_t retmaj;
     uint32_t retmin;
     char *buf;
-    char *e;
+    char *at, *sep;
 
     if (len > MAX_NAME_LEN) {
         return GSSERRS(ERR_NAMETOOLONG, GSS_S_BAD_NAME);
     }
-    buf = alloca(len + 1);
 
+    /* let's check if there is an '@' sign at all */
+    at = memchr(str, '@', len);
+    if (!at) {
+        /* not an enterprise name for sure */
+        return GSSERRS(0, GSS_S_UNAVAILABLE);
+    }
+
+    /* check if we have a '\' at all */
+    sep = memchr(str, '\\', len);
+    if (!sep) {
+        /* not an enterprise name for sure */
+        return GSSERRS(0, GSS_S_UNAVAILABLE);
+    }
+
+    /* copy buf to manipulate it */
+    buf = alloca(len + 1);
```

Comment:
Is there a problem with it ?
This is just a temporary buffer.

Reply:
I guess I can simply change it to a dynamically allocated array ... char buf[len+1]; ...

Reply:
Both of those are often prohibited in codebases due to potential for sudden very large stack allocations.  `alloca()` is often also prohibited because nothing prevents the value it returns from being returned somehow (and then becoming garbage).

Your usage is correct; I'm just surprised to see them, so I thought I'd check :)